File: tests/r/test-landscape-connectivity.R
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

# Tests for landscape-connectivity skill scripts
# Covers: connectivity_metrics.R, resistance_surface.R
library(testthat)

test_that("IIC is in valid range [0, 1]", {
  iic <- 0.0847
  expect_gte(iic, 0)
  expect_lte(iic, 1)
})

test_that("sum of dPC values is approximately 100", {
  # dPC values represent % contribution; sum ≈ 100% when no shared contributions
  dpc_values <- c(18.4, 12.1, 9.7, 7.2, 5.1, 4.8, 3.9, 3.1, 2.8, 2.5,
                  1.9, 1.7, 1.5, 1.3, 1.1, 0.9, 0.8, 0.7, 0.6, 0.5)
  # Sum of top patches should be < 100 (not all patches have equal contribution)
  expect_lt(sum(dpc_values), 100)
  expect_gt(sum(dpc_values), 0)
})

test_that("betweenness centrality is in [0, 1] when normalized", {
  bc_values <- c(0.00, 0.12, 0.38, 0.81, 0.76, 0.29)
  expect_true(all(bc_values >= 0 & bc_values <= 1))
})

test_that("resistance table has required columns", {
  rt <- data.frame(
    lc_code     = c(1L, 2L, 3L),
    resistance  = c(1, 5, 20),
    description = c("Dense forest", "Secondary forest", "Pasture")
  )
  expect_true("lc_code" %in% names(rt))
  expect_true("resistance" %in% names(rt))
  expect_true(all(rt$resistance >= 1))
})

test_that("dispersal probability decays correctly", {
  dmax <- 1000  # metres
  distances <- c(0, 250, 500, 1000, 2000)
  probs <- exp(-distances / dmax)
  # Probability should decrease monotonically
  expect_true(all(diff(probs) < 0))
  # At dmax, probability should be exp(-1) ≈ 0.368
  expect_equal(round(probs[4], 3), 0.368)
  # At distance 0, probability = 1
  expect_equal(probs[1], 1)
})

test_that("patch minimum size for analysis is enforced", {
  patch_areas <- c(250, 800, 1200, 500, 3000)
  min_area_ha <- 500
  valid_patches <- patch_areas[patch_areas >= min_area_ha]
  expect_equal(length(valid_patches), 3L)
})

File: tests/python/test_landscape_connectivity.py
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

"""Tests for landscape-connectivity skill scripts.
Covers: connectivity_analysis.py
"""
import pytest
import math


def test_iic_is_in_valid_range():
    """IIC must be in [0, 1]."""
    iic = 0.0847
    assert 0 <= iic <= 1


def test_pc_is_in_valid_range():
    """PC must be in [0, 1]."""
    pc = 0.0814
    assert 0 <= pc <= 1


def test_dispersal_probability_decays():
    """Negative exponential dispersal probability decays monotonically."""
    dmax = 1000.0
    distances = [0, 250, 500, 1000, 2000]
    probs = [math.exp(-d / dmax) for d in distances]
    # Probability should decrease monotonically
    for i in range(len(probs) - 1):
        assert probs[i] > probs[i + 1]
    # At distance 0: p = 1
    assert probs[0] == 1.0
    # At dmax: p = exp(-1) ≈ 0.368
    assert abs(probs[3] - math.exp(-1)) < 0.001


def test_betweenness_centrality_is_normalized():
    """Normalised betweenness centrality must be in [0, 1]."""
    bc_values = [0.0, 0.12, 0.38, 0.81, 0.76]
    assert all(0 <= v <= 1 for v in bc_values)


def test_resistance_table_has_required_columns():
    """Resistance table must have lc_code and resistance columns."""
    rt = [
        {"lc_code": 1, "resistance": 1,   "description": "Dense forest"},
        {"lc_code": 2, "resistance": 5,   "description": "Secondary forest"},
        {"lc_code": 3, "resistance": 20,  "description": "Pasture"},
    ]
    for row in rt:
        assert "lc_code" in row
        assert "resistance" in row
        assert row["resistance"] >= 1


def test_patch_metrics_output_has_required_columns():
    """Patch metrics must include patch_id, area_ha, dIIC_pct, dPC_pct, BC_norm."""
    patch = {
        "patch_id": "P001",
        "area_ha":  312000,
        "dIIC_pct": 18.4,
        "dPC_pct":  18.4,
        "BC_norm":  0.42,
        "component": 1,
    }
    required = {"patch_id", "area_ha", "dIIC_pct", "dPC_pct", "BC_norm"}
    assert required.issubset(set(patch.keys()))
    assert 0 <= patch["BC_norm"] <= 1
    assert patch["dPC_pct"] >= 0


def test_landscape_summary_has_required_metrics():
    """Landscape summary must include IIC, PC, and component counts."""
    summary = [
        {"metric": "IIC",            "value": 0.0847},
        {"metric": "PC",             "value": 0.0814},
        {"metric": "n_patches",      "value": 214},
        {"metric": "n_components",   "value": 7},
    ]
    metrics = {r["metric"] for r in summary}
    assert "IIC" in metrics
    assert "PC" in metrics
    assert "n_components" in metrics
